File: featuretests/innovation_filtering/nonlinear_test.cpp
/// Innovation Filtering Feature Test
///
/// Create a model with states:
/// - x, y, heading, velocity model
///
/// Provide heading readings, expect rejecting 180 degree heading errors.
/// Nonlinear model provides clear divergence signal. If innovation filtering
/// isn't working as expected, then the model will flip into the wrong
/// direction.
///
/// Passes if the model rejects the high innovation updates.

#include <featuretest/nonlinear.h>
#include <formak/runtime/ManagedFilter.h>
#include <gtest/gtest.h>

#include <random>

namespace featuretest {

namespace {
constexpr double radians(double degrees) {
  // Precision isn't important, but probably should be changed for a std math
  // operation
  return degrees / 180.0 * 3.14159;
}
constexpr double degrees(double radians) {
  return radians * 180.0 / 3.14159;
}

constexpr double TRUE_SCALE = radians(5.0);
}  // namespace

std::vector<double> rng(size_t size, size_t seed = 1) {
  std::minstd_rand prng(seed);
  std::normal_distribution<> dist(0, TRUE_SCALE);

  std::vector<double> result;

  for (size_t i = 0; i < size; i++) {
    result.push_back(dist(prng));
  }

  return result;
}

TEST(InnovationFilteringTest, ObviousInnovationRejections) {
  featuretest::State state({
      .heading = 0.0,
      .x = 1.0,
      .y = 0.0,
  });
  ASSERT_DOUBLE_EQ(state.heading(), 0.0);

  featuretest::Control control({
      ._heading_err = 0.0,
      .velocity = 1.0,
  });

  formak::runtime::ManagedFilter<featuretest::ExtendedKalmanFilter> mf(
      0.0, {
               .state = state,
               .covariance = {},
           });

  std::vector<double> readings = rng(100, 3);

  ASSERT_EQ(readings.size(), 100);
  readings[readings.size() / 4] = radians(180.0);
  readings[readings.size() / 3] = radians(180.0);
  readings[readings.size() / 2] = radians(180.0);

  for (size_t idx = 0; idx < readings.size(); ++idx) {
    auto r = readings[idx];
    auto s = mf.tick(
        idx * 0.1, control,
        {
            mf.wrap<Compass>(idx * 0.1 - 0.05, CompassOptions{.heading = r}),
        });

    ASSERT_LT(std::abs(s.state.heading()), TRUE_SCALE * 4)
        << "idx: " << idx << " reading: " << degrees(r);
  }
}

}  // namespace featuretest

File: featuretests/innovation_filtering/nonlinear_test.py
"""
Innovation Filtering Feature Test.

Create a model with states:
- x, y, heading, velocity model

Provide heading readings, expect rejecting 180 degree heading errors.
Nonlinear model provides clear divergence signal. If innovation filtering isn't
working as expected, then the model will flip into the wrong direction.

Passes if the model rejects the high innovation updates.
"""
from math import degrees, radians

import numpy as np
from sympy import cos, sin

from formak import python, runtime, ui

TRUE_SCALE = radians(5.0)


def make_ekf():
    dt = ui.Symbol("dt")

    x, y, heading, velocity, _heading_err = ui.symbols(
        ["x", "y", "heading", "velocity", "_heading_err"]
    )
    state = {x, y, heading}
    control = {velocity, _heading_err}

    state_model = {
        x: x + dt * velocity * cos(heading),
        y: y + dt * velocity * sin(heading),
        heading: heading + _heading_err,
    }

    model = ui.Model(dt=dt, state=state, control=control, state_model=state_model)

    config = python.Config(innovation_filtering=4)

    ekf = python.compile_ekf(
        symbolic_model=model,
        process_noise={velocity: 1.0, _heading_err: 0.1},
        sensor_models={"compass": {heading: heading}},
        sensor_noises={"compass": {heading: 1.0}},
        config=config,
    )
    compass_model = python.SensorModel(model, {heading: heading}, {}, config)

    return ekf, compass_model.model


def test_obvious_innovation_rejections():
    ekf, compass_model = make_ekf()

    state = ekf.State(x=1.0, y=0.0, heading=0.0)
    covariance = ekf.Covariance(x=1.0, y=1.0, heading=1.0)
    control = ekf.Control(velocity=1.0)

    mf = runtime.ManagedFilter(
        ekf=ekf, start_time=0.0, state=state, covariance=covariance
    )

    readings = np.random.default_rng(seed=3).normal(
        loc=0.0, scale=TRUE_SCALE / 2.0, size=(100,)
    )

    readings[readings.shape[0] // 4] = radians(180.0)
    readings[readings.shape[0] // 3] = radians(180.0)
    readings[readings.shape[0] // 2] = radians(180.0)

    for idx, r in enumerate(readings):
        s = mf.tick(
            0.1 * idx,
            control=control,
            readings=[runtime.StampedReading(0.1 * idx - 0.05, "compass", heading=r)],
        )

        if abs(compass_model(s.state).data) >= TRUE_SCALE * 4:
            print({"idx": idx, "reading": degrees(r)})
        assert abs(compass_model(s.state).data) < TRUE_SCALE * 4
